src/__tests__/unit/controllers/hooks/router-store/test.tsx

```diff
@@ -0,0 +1,306 @@
+      <Router routes={mockRoutes} history={history}>
+        <MockComponent>
+          {() => {
+            const [param, setParam] = useQueryParam('foo');
+            qpVal = param;
+            qpUpdateFn = setParam;
+
+            return <div>I am a subscriber</div>;
+          }}
+        </MockComponent>
+      </Router>
+    );
+
+    expect(qpVal).toEqual('hello');
+
+    act(() => qpUpdateFn('newVal'));
+
+    await nextTick();
+
+    expect(historyPushSpy).toBeCalledWith(
+      `${mockPath}?foo=newVal&bar=world#hash`
+    );
+  });
+
+  it('should remove param from URL when set to null', async () => {
+    const mockPath = mockLocation.pathname;
+    let qpVal: string | undefined, qpUpdateFn: (qp: string | null) => void;
+
+    mount(
+      <Router routes={mockRoutes} history={history}>
+        <MockComponent>
+          {() => {
+            const [param, setParam] = useQueryParam('foo');
+            qpVal = param;
+            qpUpdateFn = setParam;
+
+            return <div>I am a subscriber</div>;
+          }}
+        </MockComponent>
+      </Router>
+    );
+
+    expect(qpVal).toEqual('hello');
+
+    act(() => qpUpdateFn(null));
+
+    await nextTick();
+
+    expect(historyPushSpy).toBeCalledWith(`${mockPath}?bar=world#hash`);
+  });
+
+  it('should only re-render components hooked to a specific param', async () => {
+    let fooVal: string | undefined,
+      barVal: string | undefined,
+      fooUpdateFn: (qp: string) => void,
+      barUpdateFn: (qp: string) => void;
+
+    let renderedFoo = 0;
+    const ComponentFoo = () => {
+      const [foo, setFoo] = useQueryParam('foo');
+      fooVal = foo;
+      fooUpdateFn = setFoo;
+      renderedFoo++;
+
+      return <div id="foo-count">{renderedFoo}</div>;
+    };
+    let renderedBar = 0;
+    const ComponentBar = () => {
+      const [bar, setBar] = useQueryParam('bar');
+      barVal = bar;
+      barUpdateFn = setBar;
+      renderedBar++;
+
+      return <div id="bar-count">{renderedBar}</div>;
+    };
+
+    const wrapper = mount(
+      <Router
+        routes={mockRoutes}
+        history={historyHelper.createBrowserHistory()}
+      >
+        <ComponentFoo />
+        <ComponentBar />
+      </Router>
+    );
+
+    expect(wrapper.find(ComponentFoo).html()).toEqual(
+      '<div id="foo-count">1</div>'
+    );
+    expect(wrapper.find(ComponentBar).html()).toEqual(
+      '<div id="bar-count">1</div>'
+    );
+
+    const { actions } = getRouterStore();
+
+    actions.push('/projects/123/board/456?foo=hello&bar=world');
+    await nextTick();
+
+    expect(fooVal).toEqual('hello');
+    expect(barVal).toEqual('world');
+    expect(wrapper.find(ComponentFoo).html()).toEqual(
+      '<div id="foo-count">2</div>'
+    );
+    expect(wrapper.find(ComponentBar).html()).toEqual(
+      '<div id="bar-count">2</div>'
+    );
+
+    act(() => fooUpdateFn('newVal'));
+    await nextTick();
+
+    // URL is now — /projects/123/board/456?foo=newVal&bar=world
```

Comment:
can we assert on this? rather than doing 2 different `expects` just assert on the whole URL?

Reply:
Since we're using `createMemoryHistory()` we'll have to dig up the url from store instance. Should be able to add another assertion statement for the whole url.

Reply:
done